Group members: Tongxin Zhu, Shan Du

In [1]:
import pandas as pd
import numpy as np

# Conduct Data Access and Profiling

In [2]:
#Read the dataset
df = pd.read_csv ('~/Desktop/Passengers_Satisfaction_Survey (1)-2.csv')
df.head()

Gender      Customer Type   Age   Type of Travel     Class  \
0    Male     Loyal Customer  69.0  Personal Travel  Business   
1    Male  Disloyal Customer  20.0  Business travel       Eco   
2    Male  Disloyal Customer  41.0  Business travel  Business   
3    Male     Loyal Customer  52.0  Business travel  Business   
4  Female  Disloyal Customer  21.0  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0          -5421.0                    3.0                                2.0   
1              0.0                    3.0                                3.0   
2              0.0                    1.0                                1.0   
3              0.0                    2.0                                4.0   
4              0.0                    4.0                                0.0   

  Ease of Online booking Food and drink  Inflight entertainment  \
0                      3              5                     1.0   
1                      3              2                    10.0   
2                    NaN            NaN                     NaN   
3                      4              1                     2.0   
4                    NaN            NaN                     NaN   

  Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                1          3.0                        14.0   
1                4          2.0                         0.0   
2              NaN          NaN                         NaN   
3                2          4.0                         0.0   
4              NaN          NaN                         NaN   

   Arrival Delay in Minutes  Satisfaction  
0                       0.0  Dissatisfied  
1                       0.0  dissatisfied  
2                       NaN       neutral  
3                       6.0  Dissatisfied  
4                       NaN       neutral

In [3]:
#Subset the dataset with only 'Class' = 'Business'
df_Business = df[(df["Class"] == "Business")] 
df_Business.head()

Gender      Customer Type   Age   Type of Travel     Class  \
0    Male     Loyal Customer  69.0  Personal Travel  Business   
2    Male  Disloyal Customer  41.0  Business travel  Business   
3    Male     Loyal Customer  52.0  Business travel  Business   
4  Female  Disloyal Customer  21.0  Business travel  Business   
5  Female     Loyal Customer  57.0  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0          -5421.0                    3.0                                2.0   
2              0.0                    1.0                                1.0   
3              0.0                    2.0                                4.0   
4              0.0                    4.0                                0.0   
5              0.0                    3.0                                4.0   

  Ease of Online booking Food and drink  Inflight entertainment  \
0                      3              5                     1.0   
2                    NaN            NaN                     NaN   
3                      4              1                     2.0   
4                    NaN            NaN                     NaN   
5                      4              5                     2.0   

  Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                1          3.0                        14.0   
2              NaN          NaN                         NaN   
3                2          4.0                         0.0   
4              NaN          NaN                         NaN   
5                3          2.0                         0.0   

   Arrival Delay in Minutes  Satisfaction  
0                       0.0  Dissatisfied  
2                       NaN       neutral  
3                       6.0  Dissatisfied  
4                       NaN       neutral  
5                       0.0  Dissatisfied

In [4]:
#df_Business has 50000 rows and 16 columns
df_Business.shape

(50000, 16)

In [5]:
#The name and data type of each data column
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  ob

In [6]:
#The unique values of each column
for col in df_Business.columns:
    print(col)
    unique = df_Business[col].unique()
    print(unique, '\n*************************************\n\n')

Gender
['Male' 'Female' 'F' 'M' nan] 
*************************************


Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
*************************************


Age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 
*************************************


Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 
*************************************


Class
['Business'] 
*************************************


Flight Distance
[-5.421e+03  0.000e+00  5.600e+01 ...  4.983e+03  2.820e+06  9.000e+06] 
*************************************


Inflight wifi service
[ 3.  1.  2.  4.  0.  5. nan] 
***************************

In [7]:
#The number of missing values of each column
df_Business.isnull().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

In [8]:
#Summary statistics for each data column
df_Business.describe()

Age  Flight Distance  Inflight wifi service  \
count  47373.000000     4.660400e+04           48239.000000   
mean      41.853186     1.934737e+03               2.774519   
std       13.730683     4.369344e+04               1.423917   
min        0.000000    -5.421000e+03               0.000000   
25%       34.000000     5.940000e+02               2.000000   
50%       43.000000     1.597000e+03               3.000000   
75%       52.000000     2.565000e+03               4.000000   
max      185.000000     9.000000e+06               5.000000   

       Departure/Arrival time convenient  Inflight entertainment  \
count                       48239.000000            43649.000000   
mean                            2.910695                3.835391   
std                             1.501711                1.233508   
min                             0.000000                0.000000   
25%                             2.000000                3.000000   
50%                             3.000000                4.000000   
75%                             4.000000                5.000000   
max                             5.000000               10.000000   

        Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  
count  43649.000000                43649.000000              42098.000000  
mean       3.737520                   13.595363                 13.059647  
std        0.987664                   36.595048                 38.116530  
min        0.000000                   -2.000000                -45.000000  
25%        3.000000                    0.000000                  0.000000  
50%        4.000000                    0.000000                  0.000000  
75%        5.000000                   10.000000                 11.000000  
max        5.000000                 1305.000000               1280.000000

In [9]:
#Summary includes non numerical values
df_Business.describe(include = 'all')

Gender   Customer Type           Age   Type of Travel     Class  \
count    48239           48239  47373.000000            48239     50000   
unique       4               4           NaN                3         1   
top     Female  Loyal Customer           NaN  Business travel  Business   
freq     23001           40770           NaN            44063     50000   
mean       NaN             NaN     41.853186              NaN       NaN   
std        NaN             NaN     13.730683              NaN       NaN   
min        NaN             NaN      0.000000              NaN       NaN   
25%        NaN             NaN     34.000000              NaN       NaN   
50%        NaN             NaN     43.000000              NaN       NaN   
75%        NaN             NaN     52.000000              NaN       NaN   
max        NaN             NaN    185.000000              NaN       NaN   

        Flight Distance  Inflight wifi service  \
count      4.660400e+04           48239.000000   
unique              NaN                    NaN   
top                 NaN                    NaN   
freq                NaN                    NaN   
mean       1.934737e+03               2.774519   
std        4.369344e+04               1.423917   
min       -5.421000e+03               0.000000   
25%        5.940000e+02               2.000000   
50%        1.597000e+03               3.000000   
75%        2.565000e+03               4.000000   
max        9.000000e+06               5.000000   

        Departure/Arrival time convenient Ease of Online booking  \
count                        48239.000000                  42705   
unique                                NaN                      7   
top                                   NaN                      3   
freq                                  NaN                   8635   
mean                             2.910695                    NaN   
std                              1.501711                    NaN   
min                              0.000000                    NaN   
25%                              2.000000                    NaN   
50%                              3.000000                    NaN   
75%                              4.000000                    NaN   
max                              5.000000                    NaN   

       Food and drink  Inflight entertainment Baggage handling   Cleanliness  \
count           42705            43649.000000            43649  43649.000000   
unique              7                     NaN                6           NaN   
top                 4                     NaN                4           NaN   
freq            10890                     NaN            17056           NaN   
mean              NaN                3.835391              NaN      3.737520   
std               NaN                1.233508              NaN      0.987664   
min               NaN                0.000000              NaN      0.000000   
25%               NaN                3.000000              NaN      3.000000   
50%               NaN                4.000000              NaN      4.000000   
75%               NaN                5.000000              NaN      5.000000   
max               NaN               10.000000              NaN      5.000000   

        Departure Delay in Minutes  Arrival Delay in Minutes Satisfaction  
count                 43649.000000              42098.000000        47840  
unique                         NaN                       NaN            4  
top                            NaN                       NaN    satisfied  
freq                           NaN                       NaN        30400  
mean                     13.595363                 13.059647          NaN  
std                      36.595048                 38.116530          NaN  
min                      -2.000000                -45.000000          NaN  
25%                       0.000000                  0.000000          NaN  
50%                       0.000000          

In [10]:
#Frequency distribution
for col in df_Business.columns:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

Gender
Female    23001
Male      22830
F          1361
M          1047
Name: Gender, dtype: int64 

Customer Type
Loyal Customer       40770
Disloyal Customer     6636
disloyal               831
Loyal                    2
Name: Customer Type, dtype: int64 

Age
39.0     1935
40.0     1677
41.0     1533
42.0     1514
44.0     1506
         ... 
73.0       22
78.0       17
85.0       14
180.0       1
185.0       1
Name: Age, Length: 74, dtype: int64 

Type of Travel
Business travel    44063
Business            2122
Personal Travel     2054
Name: Type of Travel, dtype: int64 

Class
Business    50000
Name: Class, dtype: int64 

Flight Distance
337.0     225
2475.0    187
2586.0    148
404.0     133
1744.0    133
         ... 
1386.0      1
938.0       1
2780.0      1
428.0       1
768.0       1
Name: Flight Distance, Length: 3573, dtype: int64 

Inflight wifi service
2.0    10784
3.0    10593
4.0     9140
1.0     8808
5.0     7025
0.0     1889
Name: Inflight wifi service, dtype: int64 

D

In [11]:
#The dataset have total 3064 duplicated rows
duplicate = df_Business[df_Business.duplicated()]
print(df_Business.duplicated().sum())

3064


In [12]:
#Duplicated rows
duplicate

Gender      Customer Type   Age   Type of Travel     Class  \
153          F  Disloyal Customer  25.0         Business  Business   
349        NaN                NaN   NaN              NaN  Business   
369        NaN                NaN   NaN              NaN  Business   
705        NaN                NaN   NaN              NaN  Business   
1159    Female     Loyal Customer  49.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104737  Female     Loyal Customer  37.0  Business travel  Business   
104768     NaN                NaN   NaN              NaN  Business   
104789    Male     Loyal Customer  24.0         Business  Business   
104825       M     Loyal Customer  30.0  Business travel  Business   
104840       M     Loyal Customer  30.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
153                67.0                    1.0   
349                 NaN                    NaN   
369                 NaN                    NaN   
705                 NaN                    NaN   
1159              101.0                    1.0   
...                 ...                    ...   
104737           3987.0                    3.0   
104768              NaN                    NaN   
104789           3993.0                    2.0   
104825              NaN                    3.0   
104840              NaN                    4.0   

        Departure/Arrival time convenient Ease of Online booking  \
153                                   0.0                      1   
349                                   NaN                    NaN   
369                                   NaN                    NaN   
705                                   NaN                    NaN   
1159                                  4.0                      4   
...                                   ...                    ...   
104737                                1.0                      3   
104768                                NaN                    NaN   
104789                                5.0                      5   
104825                                3.0                    NaN   
104840                                4.0                    NaN   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
153                 3                     3.0                4          3.0   
349               NaN                     NaN              NaN          NaN   
369               NaN                     NaN              NaN          NaN   
705               NaN                     NaN              NaN          NaN   
1159                3                     2.0                1          3.0   
...               ...                     ...              ...          ...   
104737              5                     5.0                5          5.0   
104768            NaN                     NaN              NaN          NaN   
104789              2                     2.0                A          2.0   
104825            NaN                     4.0                1          4.0   
104840            NaN                     4.0                4          4.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
153                            0.0                       0.0           NaN  
349                            NaN                       NaN  dissatisfied  
369                            NaN                       NaN  dissatisfied  
705                            NaN                       NaN  dissatisfied  
1159                          11.0                      14.0  Dissatisfied  
...                            ...                       ...           ...  
104737                        10.0                      23.0     satisfied  
104768                         NaN                       NaN  dissatisfied  
104789                        23.0                     -45.0  dissatisfied  
104825                        

## List of Issues
1. The dataset 'df_Business' which contains all passenger satisfaction data about only the Business Class has 50000 rows and 16 columns.

2. The columns contain float and object values.

3. There are improperly formatted data in almost all columns. Non numerial columns have duplicated values, and numerial columns have some negative values and some other values do not make sense.

4. The observed unique values are shown above.

5. Only 'Class' column does not have missing values since the dataset was filtered by specific 'Business Class', whereas other columns all have missing values.

6. Summary statistics have been shown above. 

7. Frequency distribution has been shown above.

8. There are total 3064 duplicated rows.

# Conduct Data Cleansing

In [13]:
#Drop the Class column
df_Business.drop(['Class'], axis = 1, inplace = False)
df_Business = df_Business.reset_index(drop = True)
df_Business.head()

Gender      Customer Type   Age   Type of Travel     Class  \
0    Male     Loyal Customer  69.0  Personal Travel  Business   
1    Male  Disloyal Customer  41.0  Business travel  Business   
2    Male     Loyal Customer  52.0  Business travel  Business   
3  Female  Disloyal Customer  21.0  Business travel  Business   
4  Female     Loyal Customer  57.0  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0          -5421.0                    3.0                                2.0   
1              0.0                    1.0                                1.0   
2              0.0                    2.0                                4.0   
3              0.0                    4.0                                0.0   
4              0.0                    3.0                                4.0   

  Ease of Online booking Food and drink  Inflight entertainment  \
0                      3              5                     1.0   
1                    NaN            NaN                     NaN   
2                      4              1                     2.0   
3                    NaN            NaN                     NaN   
4                      4              5                     2.0   

  Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                1          3.0                        14.0   
1              NaN          NaN                         NaN   
2                2          4.0                         0.0   
3              NaN          NaN                         NaN   
4                3          2.0                         0.0   

   Arrival Delay in Minutes  Satisfaction  
0                       0.0  Dissatisfied  
1                       NaN       neutral  
2                       6.0  Dissatisfied  
3                       NaN       neutral  
4                       0.0  Dissatisfied

In [14]:
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Food and drink', 'Inflight entertainment', 'Baggage handling',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'Satisfaction'],
      dtype='object')

In [15]:
#Rename columns
df_Business.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)
df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'class',
       'flight_distance', 'inflight_wifi_service',
       'departure/arrival_time_convenient', 'ease_of_online_booking',
       'food_and_drink', 'inflight_entertainment', 'baggage_handling',
       'cleanliness', 'departure_delay_in_minutes', 'arrival_delay_in_minutes',
       'satisfaction'],
      dtype='object')

### Modify improperly formatted data and handle bad data
Base on the dsta dictionary provided

In [16]:
#Fix the gender column
#Replace 'F', 'M' with 'Female', 'Male'
df_Business['gender'] = df_Business['gender'].replace(['F', 'M'],['Female', 'Male'])

In [17]:
#Check updated gender column
df_Business['gender'].unique()

array(['Male', 'Female', nan], dtype=object)

In [18]:
#Fix the customer_type column
#Replace 'disloyal' with 'Disloyal Customer' and 'loyal', 'Loyal' with 'Loyal Customer'
df_Business['customer_type'] = df_Business['customer_type'].replace(['disloyal', 'loyal', 'Loyal'],['Disloyal Customer', 'Loyal Customer', 'Loyal Customer'])

In [19]:
#Check updated customer_type column
df_Business['customer_type'].unique()

array(['Loyal Customer', 'Disloyal Customer', nan], dtype=object)

In [20]:
#Fix the age column
#Replace 180 with 80, and 185 with 85 
#Replace 0 with the first mode value 
df_Business['age'] = df_Business['age'].replace([180, 185],[80, 85])
mode_value = df_Business['age'].mode()
df_Business['age'] = df_Business['age'].replace([0],[mode_value[0]])

In [21]:
#Check updated age column
df_Business['age'].unique()

array([69., 41., 52., 21., 57., 45., 39., 33., 10., 55., 59., 53., 56.,
       47., 54., 25., 51., 60., 27., 43., 20., 19., 63., 50., nan, 38.,
       46., 85., 26., 42., 30., 24., 22., 17., 48., 40., 16., 32., 35.,
       44., 49., 65., 67., 66., 64.,  7., 37., 36., 15., 62., 34., 68.,
       70., 58., 11., 74., 13., 80., 75.,  9., 61., 71., 12., 73., 72.,
       18., 79.,  8., 76., 14., 77., 78.])

In [22]:
#Fix the type_of_travel column 
#Replace 'Business', 'Business travel' with "Business Travel"
df_Business['type_of_travel'] = df_Business['type_of_travel'].replace(['Business', 'Business travel'],['Business Travel', 'Business Travel'])

In [23]:
#Check updated type_of_travel column
df_Business['type_of_travel'].unique()

array(['Personal Travel', 'Business Travel', nan], dtype=object)

In [24]:
#Fix the flight_distance column 
#Replace negative value with a positive one
#Replace 0 with the first mode value
df_Business['flight_distance'] = df_Business['flight_distance'].replace([-5.421e+03],[5.421e+03])
mode_value = df_Business['flight_distance'].mode()
df_Business['flight_distance'] = df_Business['flight_distance'].replace([0.000e+00],[mode_value[0]])

In [26]:
#Check updated flight_distance column
df_Business['flight_distance'].unique()
sorted(df_Business['flight_distance'].unique())

[56.0,
 67.0,
 73.0,
 74.0,
 76.0,
 77.0,
 78.0,
 80.0,
 82.0,
 83.0,
 84.0,
 86.0,
 89.0,
 90.0,
 95.0,
 96.0,
 98.0,
 100.0,
 101.0,
 102.0,
 106.0,
 107.0,
 108.0,
 109.0,
 110.0,
 112.0,
 113.0,
 114.0,
 115.0,
 116.0,
 118.0,
 119.0,
 120.0,
 121.0,
 122.0,
 123.0,
 125.0,
 126.0,
 127.0,
 128.0,
 129.0,
 130.0,
 133.0,
 134.0,
 135.0,
 137.0,
 139.0,
 140.0,
 141.0,
 142.0,
 143.0,
 144.0,
 145.0,
 337.0,
 5421.0,
 nan,
 146.0,
 147.0,
 148.0,
 150.0,
 151.0,
 152.0,
 153.0,
 154.0,
 156.0,
 157.0,
 158.0,
 159.0,
 160.0,
 162.0,
 163.0,
 164.0,
 166.0,
 168.0,
 169.0,
 170.0,
 171.0,
 173.0,
 174.0,
 175.0,
 177.0,
 179.0,
 180.0,
 181.0,
 182.0,
 183.0,
 184.0,
 185.0,
 187.0,
 188.0,
 189.0,
 190.0,
 191.0,
 192.0,
 193.0,
 194.0,
 196.0,
 197.0,
 198.0,
 199.0,
 200.0,
 201.0,
 202.0,
 203.0,
 204.0,
 205.0,
 206.0,
 207.0,
 208.0,
 209.0,
 210.0,
 211.0,
 212.0,
 213.0,
 214.0,
 215.0,
 216.0,
 217.0,
 218.0,
 220.0,
 221.0,
 222.0,
 223.0,
 224.0,
 225.0,
 226.0,
 227.0,
 2

In [27]:
#Check unique values of inflight_wifi_service column
df_Business['inflight_wifi_service'].unique()

array([ 3.,  1.,  2.,  4.,  0.,  5., nan])

In [28]:
#Fix the departure/arrival_time_convenient column
#Replace '0' with '1' becase we assume 0 indicates 'not good'
df_Business['departure/arrival_time_convenient'] = df_Business['departure/arrival_time_convenient'].replace([0],[1])

In [29]:
#Check updated departure/arrival_time_convenient column
df_Business['departure/arrival_time_convenient'].unique()

array([ 2.,  1.,  4.,  5.,  3., nan])

In [30]:
#Fix the ease_of_online_booking column
#Replace '0' with '1' and 'Unknown' with the first mode value
mode_value = df_Business['ease_of_online_booking'].mode()
df_Business['ease_of_online_booking'] = df_Business['ease_of_online_booking'].replace(['0', 'Unknown'],['1', mode_value[0]])

In [31]:
#Check updated ease_of_online_booking column
df_Business['ease_of_online_booking'].unique()

array(['3', nan, '4', '1', '5', '2'], dtype=object)

In [32]:
#Fix the food_and_drink column
#Replace 'Great' with '5' because 5 is the highest score which can be identified as great, and '0' with '1'
#Also, to unify the data format, replace int 5 with string '5' because there are still some null values in the column
#We will convert the data type later
mode_value = df_Business['food_and_drink'].mode()
df_Business['food_and_drink'] = df_Business['food_and_drink'].replace(['Great', 5, '0'],['5', '5', '1'])

In [33]:
#Check updated food_and_drink' column
df_Business['food_and_drink'].unique()

array(['5', nan, '1', '3', '4', '2'], dtype=object)

In [34]:
#Fix the inflight_entertainment column
#Replace '0' with '1', and '10' with '5' because 10 is a high score
#Some poeple might think the highest score is 10
mode_value = df_Business['inflight_entertainment'].mode()
df_Business['inflight_entertainment'] = df_Business['inflight_entertainment'].replace([0, 10],[1, 5])

In [35]:
#Check updated inflight_entertainment column
df_Business['inflight_entertainment'].unique()

array([ 1., nan,  2.,  3.,  4.,  5.])

In [36]:
#Fix the baggage_handling column
#Replace 'A' with '5' since A is always idtified as a good score, and unify the data format. 
df_Business['baggage_handling'] = df_Business['baggage_handling'].replace(['A', 5],['5', '5'])

In [37]:
#Check updated baggage_handling column
df_Business['baggage_handling'].unique()

array(['1', nan, '2', '3', '5', '4'], dtype=object)

In [38]:
#Fix the cleanliness column
#Replace '0.' with '1'
df_Business['cleanliness'] = df_Business['cleanliness'].replace([0.],[1])

In [39]:
#Check updated cleanliness column
df_Business['cleanliness'].unique()

array([ 3., nan,  4.,  2.,  5.,  1.])

In [40]:
#Fix the departure_delay_in_minutes column
#Replace the negative value with a positive one
df_Business['departure_delay_in_minutes'] = df_Business['departure_delay_in_minutes'].replace([-2.000e+00], [2.000e+00])

In [41]:
#Check updated departure_delay_in_minutes column
sorted(df_Business['departure_delay_in_minutes'].unique())

[4.0,
 12.0,
 14.0,
 nan,
 0.0,
 1.0,
 2.0,
 3.0,
 5.0,
 6.0,
 7.0,
 8.0,
 9.0,
 10.0,
 11.0,
 13.0,
 15.0,
 16.0,
 17.0,
 18.0,
 19.0,
 20.0,
 21.0,
 22.0,
 23.0,
 24.0,
 25.0,
 26.0,
 27.0,
 28.0,
 29.0,
 30.0,
 31.0,
 32.0,
 33.0,
 34.0,
 35.0,
 36.0,
 37.0,
 38.0,
 39.0,
 40.0,
 41.0,
 42.0,
 43.0,
 44.0,
 45.0,
 46.0,
 47.0,
 48.0,
 49.0,
 50.0,
 51.0,
 52.0,
 53.0,
 54.0,
 55.0,
 56.0,
 57.0,
 58.0,
 59.0,
 60.0,
 61.0,
 62.0,
 63.0,
 64.0,
 65.0,
 66.0,
 67.0,
 68.0,
 69.0,
 70.0,
 71.0,
 72.0,
 73.0,
 74.0,
 75.0,
 76.0,
 77.0,
 78.0,
 79.0,
 80.0,
 81.0,
 82.0,
 83.0,
 84.0,
 85.0,
 86.0,
 87.0,
 88.0,
 89.0,
 90.0,
 91.0,
 92.0,
 93.0,
 94.0,
 95.0,
 96.0,
 97.0,
 98.0,
 99.0,
 100.0,
 101.0,
 102.0,
 103.0,
 104.0,
 105.0,
 106.0,
 107.0,
 108.0,
 109.0,
 110.0,
 111.0,
 112.0,
 113.0,
 114.0,
 115.0,
 116.0,
 117.0,
 118.0,
 119.0,
 120.0,
 121.0,
 122.0,
 123.0,
 124.0,
 125.0,
 126.0,
 127.0,
 128.0,
 129.0,
 130.0,
 131.0,
 132.0,
 133.0,
 134.0,
 135.0,
 136.0,
 137.0,


In [42]:
#Fix the arrival_delay_in_minutes column
#Replace the negative value with a positive one
df_Business['arrival_delay_in_minutes'] = df_Business['arrival_delay_in_minutes'].replace([-4.500e+01], [4.500e+01])

In [43]:
#Check updated arrival_delay_in_minutes column
sorted(df_Business['arrival_delay_in_minutes'].unique())

[0.0,
 nan,
 1.0,
 2.0,
 3.0,
 4.0,
 5.0,
 6.0,
 7.0,
 8.0,
 9.0,
 10.0,
 11.0,
 12.0,
 13.0,
 14.0,
 15.0,
 16.0,
 17.0,
 18.0,
 19.0,
 20.0,
 21.0,
 22.0,
 23.0,
 24.0,
 25.0,
 26.0,
 27.0,
 28.0,
 29.0,
 30.0,
 31.0,
 32.0,
 33.0,
 34.0,
 35.0,
 36.0,
 37.0,
 38.0,
 39.0,
 40.0,
 41.0,
 42.0,
 43.0,
 44.0,
 45.0,
 46.0,
 47.0,
 48.0,
 49.0,
 50.0,
 51.0,
 52.0,
 53.0,
 54.0,
 55.0,
 56.0,
 57.0,
 58.0,
 59.0,
 60.0,
 61.0,
 62.0,
 63.0,
 64.0,
 65.0,
 66.0,
 67.0,
 68.0,
 69.0,
 70.0,
 71.0,
 72.0,
 73.0,
 74.0,
 75.0,
 76.0,
 77.0,
 78.0,
 79.0,
 80.0,
 81.0,
 82.0,
 83.0,
 84.0,
 85.0,
 86.0,
 87.0,
 88.0,
 89.0,
 90.0,
 91.0,
 92.0,
 93.0,
 94.0,
 95.0,
 96.0,
 97.0,
 98.0,
 99.0,
 100.0,
 101.0,
 102.0,
 103.0,
 104.0,
 105.0,
 106.0,
 107.0,
 108.0,
 109.0,
 110.0,
 111.0,
 112.0,
 113.0,
 114.0,
 115.0,
 116.0,
 117.0,
 118.0,
 119.0,
 120.0,
 121.0,
 122.0,
 123.0,
 124.0,
 125.0,
 126.0,
 127.0,
 128.0,
 129.0,
 130.0,
 131.0,
 132.0,
 133.0,
 134.0,
 135.0,
 136.0,
 137.0,


In [44]:
#Fix the satisfaction column
#Replkace 'satisfied' with 'Satisfied', 'dissatisfied' with 'Dissatisfied', and 'neutral' with 'Neutral'
df_Business['satisfaction'] = df_Business['satisfaction'].replace(['satisfied', 'dissatisfied', 'neutral'], ['Satisfied', 'Dissatisfied', 'Neutral'])

In [45]:
#Check updated satisfaction column
df_Business['satisfaction'].unique()

array(['Dissatisfied', 'Neutral', 'Satisfied', nan], dtype=object)

In [46]:
#Check updated dataset
df_Business

gender      customer_type   age   type_of_travel     class  \
0        Male     Loyal Customer  69.0  Personal Travel  Business   
1        Male  Disloyal Customer  41.0  Business Travel  Business   
2        Male     Loyal Customer  52.0  Business Travel  Business   
3      Female  Disloyal Customer  21.0  Business Travel  Business   
4      Female     Loyal Customer  57.0  Business Travel  Business   
...       ...                ...   ...              ...       ...   
49995    Male     Loyal Customer  45.0  Business Travel  Business   
49996  Female     Loyal Customer  35.0  Business Travel  Business   
49997    Male     Loyal Customer  47.0  Business Travel  Business   
49998    Male     Loyal Customer  42.0  Business Travel  Business   
49999  Female  Disloyal Customer  39.0  Business Travel  Business   

       flight_distance  inflight_wifi_service  \
0               5421.0                    3.0   
1                337.0                    1.0   
2                337.0                    2.0   
3                337.0                    4.0   
4                337.0                    3.0   
...                ...                    ...   
49995           3305.0                    2.0   
49996            337.0                    5.0   
49997           3496.0                    2.0   
49998           3605.0                    4.0   
49999            862.0                    5.0   

       departure/arrival_time_convenient ease_of_online_booking  \
0                                    2.0                      3   
1                                    1.0                    NaN   
2                                    4.0                      4   
3                                    1.0                    NaN   
4                                    4.0                      4   
...                                  ...                    ...   
49995                                2.0                      2   
49996                                5.0                      5   
49997                                2.0                      2   
49998                                4.0                      4   
49999                                5.0                      5   

      food_and_drink  inflight_entertainment baggage_handling  cleanliness  \
0                  5                     1.0                1          3.0   
1                NaN                     NaN              NaN          NaN   
2                  1                     2.0                2          4.0   
3                NaN                     NaN              NaN          NaN   
4                  5                     2.0                3          2.0   
...              ...                     ...              ...          ...   
49995              3                     4.0                4          3.0   
49996              2                     4.0                4          5.0   
49997              2                     5.0                5          4.0   
49998              3                     4.0                4          3.0   
49999              5                     5.0                1          5.0   

       departure_delay_in_minutes  arrival_delay_in_minutes  satisfaction  
0                            14.0                       0.0  Dissatisfied  
1                             NaN                       NaN       Neutral  
2                             0.0                       6.0  Dissatisfied  
3                             NaN                       NaN       Neutral  
4                             0.0                       0.0  Dissatisfied  
...                           ...                       ...           ...  
49995                         8.0                       0.0     Satisfied  
49996                        54.0                      40.0     Satisfied  
49997                         0.0                       2.0     Satisfied  
49998                         0.0                       0.0     Satisfied  
49999      

In [47]:
#Missing values
df_Business.isnull().sum()

gender                               1761
customer_type                        1761
age                                  2627
type_of_travel                       1761
class                                   0
flight_distance                      3396
inflight_wifi_service                1761
departure/arrival_time_convenient    1761
ease_of_online_booking               7295
food_and_drink                       7295
inflight_entertainment               6351
baggage_handling                     6351
cleanliness                          6351
departure_delay_in_minutes           6351
arrival_delay_in_minutes             7902
satisfaction                         2160
dtype: int64

In [48]:
#Fill missing values of categorical columns with 'Unknown' and numerical columns with the first mode value of each column
df_Business['gender'].fillna('Unknown', inplace = True)
df_Business['customer_type'].fillna('Unknown', inplace = True)
df_Business['age'].fillna(df_Business['age'].mode()[0], inplace = True)
df_Business['type_of_travel'].fillna('Unknown', inplace = True)
df_Business['flight_distance'].fillna(df_Business['flight_distance'].mode()[0], inplace = True)
df_Business['inflight_wifi_service'].fillna(df_Business['inflight_wifi_service'].mode()[0], inplace = True)
df_Business['departure/arrival_time_convenient'].fillna(df_Business['departure/arrival_time_convenient'].mode()[0], inplace = True)
df_Business['ease_of_online_booking'].fillna(df_Business['ease_of_online_booking'].mode()[0], inplace = True)
df_Business['food_and_drink'].fillna(df_Business['food_and_drink'].mode()[0], inplace = True)
df_Business['inflight_entertainment'].fillna(df_Business['inflight_entertainment'].mode()[0], inplace = True)
df_Business['baggage_handling'].fillna(df_Business['baggage_handling'].mode()[0], inplace = True)
df_Business['cleanliness'].fillna(df_Business['cleanliness'].mode()[0], inplace = True)
df_Business['departure_delay_in_minutes'].fillna(df_Business['departure_delay_in_minutes'].mode()[0], inplace = True)
df_Business['arrival_delay_in_minutes'].fillna(df_Business['arrival_delay_in_minutes'].mode()[0], inplace = True)
df_Business['satisfaction'].fillna('Unknown', inplace = True)

In [49]:
#Check if there are still missing values
df_Business.isnull().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
class                                0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
baggage_handling                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64

In [50]:
#Remove the fully duplicated observations
df_Business.drop_duplicates(inplace = True)

In [51]:
#Check if there are still duplicated rows
df_Business.duplicated().sum()

0

In [52]:
#Fix the columns datatypes
#Covert 'gender', 'customer_type', 'type_of_travel', 'class', 'satisfaction' to string
#Covert 'flight_distance', 'departure_delay_in_minutes', 'arrival_delay_in_minutes' to float
#Covert rest columns to int
df_Business['age'] = df_Business['age'].astype("Int64")
df_Business['inflight_wifi_service'] = df_Business['inflight_wifi_service'].astype("Int64")
df_Business['departure/arrival_time_convenient'] = df_Business['departure/arrival_time_convenient'].astype("Int64")
df_Business['ease_of_online_booking'] = df_Business['ease_of_online_booking'].astype('float').astype('Int64')
df_Business['food_and_drink'] = df_Business['food_and_drink'].astype('float').astype('Int64')
df_Business['inflight_entertainment'] = df_Business['inflight_entertainment'].astype('float').astype('Int64')
df_Business['baggage_handling'] = df_Business['baggage_handling'].astype('float').astype('Int64')
df_Business['cleanliness'] = df_Business['cleanliness'].astype('float').astype('Int64')
df_Business['gender'] = df_Business['gender'].astype("string")
df_Business['customer_type'] = df_Business['customer_type'].astype("string")
df_Business['type_of_travel'] = df_Business['type_of_travel'].astype("string")
df_Business['class'] = df_Business['class'].astype("string")
df_Business['satisfaction'] = df_Business['satisfaction'].astype("string")

In [53]:
#Updated data types
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46931 entries, 0 to 49999
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             46931 non-null  string 
 1   customer_type                      46931 non-null  string 
 2   age                                46931 non-null  Int64  
 3   type_of_travel                     46931 non-null  string 
 4   class                              46931 non-null  string 
 5   flight_distance                    46931 non-null  float64
 6   inflight_wifi_service              46931 non-null  Int64  
 7   departure/arrival_time_convenient  46931 non-null  Int64  
 8   ease_of_online_booking             46931 non-null  Int64  
 9   food_and_drink                     46931 non-null  Int64  
 10  inflight_entertainment             46931 non-null  Int64  
 11  baggage_handling                   46931 non-null  Int

In [54]:
#Final version of dataset
df_Business

gender      customer_type  age   type_of_travel     class  \
0        Male     Loyal Customer   69  Personal Travel  Business   
1        Male  Disloyal Customer   41  Business Travel  Business   
2        Male     Loyal Customer   52  Business Travel  Business   
3      Female  Disloyal Customer   21  Business Travel  Business   
4      Female     Loyal Customer   57  Business Travel  Business   
...       ...                ...  ...              ...       ...   
49995    Male     Loyal Customer   45  Business Travel  Business   
49996  Female     Loyal Customer   35  Business Travel  Business   
49997    Male     Loyal Customer   47  Business Travel  Business   
49998    Male     Loyal Customer   42  Business Travel  Business   
49999  Female  Disloyal Customer   39  Business Travel  Business   

       flight_distance  inflight_wifi_service  \
0               5421.0                      3   
1                337.0                      1   
2                337.0                      2   
3                337.0                      4   
4                337.0                      3   
...                ...                    ...   
49995           3305.0                      2   
49996            337.0                      5   
49997           3496.0                      2   
49998           3605.0                      4   
49999            862.0                      5   

       departure/arrival_time_convenient  ease_of_online_booking  \
0                                      2                       3   
1                                      1                       3   
2                                      4                       4   
3                                      1                       3   
4                                      4                       4   
...                                  ...                     ...   
49995                                  2                       2   
49996                                  5                       5   
49997                                  2                       2   
49998                                  4                       4   
49999                                  5                       5   

       food_and_drink  inflight_entertainment  baggage_handling  cleanliness  \
0                   5                       1                 1            3   
1                   4                       4                 4            4   
2                   1                       2                 2            4   
3                   4                       4                 4            4   
4                   5                       2                 3            2   
...               ...                     ...               ...          ...   
49995               3                       4                 4            3   
49996               2                       4                 4            5   
49997               2                       5                 5            4   
49998               3                       4                 4            3   
49999               5                       5                 1            5   

       departure_delay_in_minutes  arrival_delay_in_minutes  satisfaction  
0                            14.0                       0.0  Dissatisfied  
1                             0.0                       0.0       Neutral  
2                             0.0                       6.0  Dissatisfied  
3                             0.0                       0.0       Neutral  
4                             0.0                       0.0  Dissatisfied  
...                           ...                       ...           ...  
49995                         8.0                       0.0     Satisfied  
49996                        54.0                      40.0     Satisfied  
49997                         0.0                       2.0     Satisfied  
49998                         0.0                       0.0    

# Export the cleaned dataset to CSV file

In [55]:
#Exporting data into csv file
df_Business.to_csv("~/Downloads/Project_2_Group1.csv", index= False)

# Compare the Data Characteristics Before and After Data Cleansing

In [56]:
#Shape of new dataset
df_Business.shape

(46931, 16)

In [57]:
#Unique values of new dataset
for col in df_Business.columns:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

gender
Female     24058
Male       22871
Unknown        2
Name: gender, dtype: Int64 

customer_type
Loyal Customer       39655
Disloyal Customer     7274
Unknown                  2
Name: customer_type, dtype: Int64 

age
39    3590
40    1659
41    1511
42    1501
44    1491
      ... 
74      26
79      23
73      22
78      15
85      15
Name: age, Length: 71, dtype: Int64 

type_of_travel
Business Travel    44904
Personal Travel     2025
Unknown                2
Name: type_of_travel, dtype: Int64 

class
Business    46931
Name: class, dtype: Int64 

flight_distance
337.0     1056
2475.0     183
2586.0     147
404.0      133
1744.0     132
          ... 
428.0        1
218.0        1
1226.0       1
245.0        1
923.0        1
Name: flight_distance, Length: 3572, dtype: int64 

inflight_wifi_service
2    10509
3    10298
4     8882
1     8542
5     6846
0     1854
Name: inflight_wifi_service, dtype: Int64 

departure/arrival_time_convenient
1    10391
4     9697
3     8953
5     89

## Changes in each column
Total data points were reduced from 50000 to 46931 as we removed duplicated rows. <br>
**Gender** <br>
New name: gender <br>
Number of missing vales: 1761 -> 0 <br>
Filled missing values with 'Unknown' <br>
Data type: object -> string <br>
Unique values: 'Female', 'Male', 'F', 'M' -> 'Female', 'Male' <br>
**Customer Type** <br>
New name: customer_type <br>
Number of missing vales: 1761 -> 0 <br>
Filled missing values with 'Unknown' <br>
Data type: object -> string <br>
Unique values: 'Loyal Customer', 'Disloyal Customer', 'disloyal', 'Loyal' -> 'Loyal Customer', 'Disloyal Customer' <br>
**Age** <br>
New name: age <br>
Number of missing vales: 2627 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> int <br>
Unique values show below <br>
**Type of Travel** <br>
New name: type_of_travel <br>
Number of missing vales: 1761 -> 0 <br>
Filled missing values with 'Unknown' <br>
Data type: object -> string <br>
Unique values: 'Business travel', 'Business', 'Personal Travel' -> 'Business Travel', 'Personal Travel'<br>
**Class** <br>
New name: class <br>
Data type: object -> string <br>
**Flight Distance** <br>
New name: flight_distance <br>
Number of missing vales: 3396 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> float <br>
Unique values show below<br>
**Inflight wifi service** <br>
New name: inflight_wifi_service <br>
Number of missing vales: 1761 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> int <br>
Changed unique values to 0-5 <br>
**Departure/Arrival time convenient** <br>
New name: departure/arrival_time_convenient <br>
Number of missing vales: 1761 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> int <br>
Changed unique values to 1-5 <br>
**Ease of Online booking** <br>
New name: ease_of_online_booking <br>
Number of missing vales: 7295 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: object -> int <br>
Changed unique values to 1-5 <br>
**Food and drink** <br>
New name: food_and_drink <br>
Number of missing vales: 7295 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: object -> int <br>
Changed unique values to 1-5 <br>
**Inflight entertainment** <br>
New name: inflight_entertainment <br>
Number of missing vales: 6351 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> int <br>
Changed unique values to 1-5 <br>
**Baggage handling** <br>
New name: baggage_handling <br>
Number of missing vales: 6351 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: object -> int <br>
Changed unique values to 1-5 <br>
**Cleanliness** <br>
New name: cleanliness <br>
Number of missing vales: 6351 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> int <br>
Changed unique values to 1-5 <br>
**Departure Delay in Minutes** <br>
New name: departure_delay_in_minutes <br>
Number of missing vales: 6351 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> float <br>
Unique values show below <br>
**Arrival Delay in Minutes** <br>
New name: arrival_delay_in_minutes <br>
Number of missing vales: 7901 -> 0 <br>
Filled missing values with the first mode value, Replaced bad data <br>
Data type: float -> float <br>
Unique values show below <br>
**Satisfaction** <br>
New name: satisfaction <br>
Number of missing vales: 2160 -> 0 <br>
Filled missing values with 'Unknown' <br>
Data type: object -> string <br>
Unique values: 'satisfies', 'dissatisfied', 'neutral' -> 'Satisfied', 'Dissatisfied', 'Neutral' <br>

## Columns name and data type before cleaning

## Columns name and data type after cleaning

## Unique values and frequency distribution before cleaning

## Unique values and frequency distribution after cleaning

## Number of missing values before cleaning

## Number of missing values after cleaning